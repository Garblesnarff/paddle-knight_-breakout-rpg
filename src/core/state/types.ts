import { GameStatus } from '@/types';

export interface UIState {
  gameStatus: GameStatus;
  selectedWorldId: number | null;
  selectedStageId: number | null;
}

export interface RootState {
  ui: UIState;
}


