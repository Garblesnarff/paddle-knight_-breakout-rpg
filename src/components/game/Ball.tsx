import React from 'react';
import { Ball as BallType, Cosmetics, RunicEmpowermentBuffs } from '../../types/game.types';

interface BallProps {
  ball: BallType;
  cosmetics: Cosmetics;
  activeBuffs: RunicEmpowermentBuffs;
}

export const Ball: React.FC<BallProps> = ({ ball, cosmetics, activeBuffs }) => {
  const ballStyle: React.CSSProperties = {
    left: ball.x,
    top: ball.y,
    width: ball.size * 2,
    height: ball.size * 2,
  };

  if (ball.slowedUntil && Date.now() < ball.slowedUntil) {
    ballStyle.filter = 'brightness(0.7) saturate(1.5)';
    ballStyle.boxShadow = '0 0 8px 2px #67e8f9';
  }
  if (ball.isSpikeSlowedUntil && Date.now() < ball.isSpikeSlowedUntil) {
    ballStyle.filter = 'brightness(0.5) saturate(2)';
    ballStyle.boxShadow = '0 0 12px 4px #93c5fd';
  }
  if (activeBuffs.power) {
    ballStyle.boxShadow = '0 0 10px 4px #f87171';
  }
  if (activeBuffs.haste) {
    ballStyle.boxShadow = '0 0 10px 4px #38bdf8';
  }
  if (cosmetics.ballEffect === 'magical') {
    ballStyle.boxShadow = (ballStyle.boxShadow ? ballStyle.boxShadow + ', ' : '') + '0 0 15px 5px #a855f7';
  }

  return (
    <div
      className="absolute rounded-full bg-gradient-to-br from-yellow-300 to-orange-500 shadow-xl border-2 border-yellow-100/50"
      style={ballStyle}
    ></div>
  );
};
