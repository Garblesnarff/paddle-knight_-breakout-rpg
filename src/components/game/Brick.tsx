import React from 'react';
import { Brick as BrickType, BrickType as BrickEnum } from '../../types/game.types';
import { BRICK_PROPERTIES, BOSS_ENRAGE_THRESHOLD } from '../../constants/game.constants';

interface BrickProps {
  brick: BrickType;
}

export const Brick: React.FC<BrickProps> = ({ brick }) => {
  const brickProps = BRICK_PROPERTIES[brick.type];
  const hpPercentage = (brick.hp / brick.maxHp) * 100;
  let isEnraged = brick.type === BrickEnum.Boss && hpPercentage <= BOSS_ENRAGE_THRESHOLD;
  if (brick.type === BrickEnum.ArchmageBoss && (brick.phase === 2 || brick.phase === 3)) {
    isEnraged = true;
  }

  let brickColor = isEnraged ? 'bg-red-600' : brickProps.color;

  if (brick.type === BrickEnum.Chaos) {
    brickColor = 'bg-gradient-to-br from-purple-600 via-pink-600 to-red-700';
  } else if (brick.type === BrickEnum.Mirror) {
    brickColor = 'bg-gradient-to-br from-slate-100 to-slate-400';
  } else if (brick.type === BrickEnum.Rune) {
    brickColor = 'bg-gradient-to-br from-purple-600 to-indigo-800';
  }

  const brickClasses = `absolute rounded-md shadow-inner shadow-black/40 ${brickColor} border-t-2 border-l-2 border-gray-500/30 border-b-2 border-r-2 border-black/30`;
  const brickStyle: React.CSSProperties = {
    left: brick.x,
    top: brick.y,
    width: brick.width,
    height: brick.height,
    transition: 'opacity 0.3s',
  };

  if (brick.isClone) {
    brickStyle.opacity = 0.6;
    brickStyle.filter = 'saturate(0.5)';
  }

  // Shield visual takes precedence
  if (brick.shieldHp && brick.shieldHp > 0) {
    brickStyle.boxShadow = '0 0 12px 5px #67e8f9'; // Bright cyan glow
    brickStyle.border = '1px solid #cffafe';
  } else if (brick.type === BrickEnum.Apprentice) {
    Object.assign(brickStyle, { boxShadow: '0 0 8px 2px #38bdf8' });
  } else if (brick.type === BrickEnum.Ice) {
    Object.assign(brickStyle, { boxShadow: '0 0 10px 3px #67e8f9' });
  } else if (brick.type === BrickEnum.Lightning) {
    Object.assign(brickStyle, { boxShadow: '0 0 12px 3px #facc15' });
  } else if (brick.type === BrickEnum.Mirror) {
    Object.assign(brickStyle, { boxShadow: '0 0 15px 4px #e2e8f0' });
  }

  if (isEnraged) {
    brickStyle.animation = 'pulse 1s cubic-bezier(0.4, 0, 0.6, 1) infinite';
    brickStyle.boxShadow = '0 0 20px 8px #a855f7';
  }

  return (
    <div style={brickStyle} className={brickClasses}>
      {brick.type === BrickEnum.Rune && (
        <span className="text-xl text-yellow-300 font-medieval absolute inset-0 flex items-center justify-center select-none opacity-80">
          ‚ùñ
        </span>
      )}
      {brick.maxHp > 1 && !brick.isClone && (
        <div className="absolute -bottom-1 w-full h-1 bg-gray-900/50 rounded-full p-px">
          <div
            className="bg-green-500 h-full rounded-full"
            style={{ width: `${hpPercentage}%` }}
          ></div>
        </div>
      )}
    </div>
  );
};
