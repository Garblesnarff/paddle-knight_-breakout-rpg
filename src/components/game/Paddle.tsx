import React from 'react';
import { PADDLE_Y, PADDLE_HEIGHT } from '../../constants/game.constants';

interface PaddleProps {
  paddleX: number;
  paddleWidth: number;
}

export const Paddle: React.FC<PaddleProps> = ({ paddleX, paddleWidth }) => {
  const paddleStyle: React.CSSProperties = {
    left: paddleX,
    top: PADDLE_Y,
    width: paddleWidth,
    height: PADDLE_HEIGHT,
  };

  return (
    <div
      className="absolute bg-gradient-to-b from-slate-400 to-slate-600 rounded-sm shadow-lg border-t-2 border-slate-300 border-b-2 border-slate-700"
      style={paddleStyle}
    ></div>
  );
};
