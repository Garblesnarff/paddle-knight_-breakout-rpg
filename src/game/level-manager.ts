import { Brick, BrickType } from '../types';
import { GAME_WIDTH, BRICK_WIDTH, BRICK_HEIGHT, BRICK_GAP, BRICK_PROPERTIES, BOSS_MOVE_SPEED } from '../constants';
import { LEVEL_LAYOUTS as STAGE1_LAYOUTS } from './stages/stage-1/layouts';
import { LEVEL_LAYOUTS as STAGE2_LAYOUTS } from './stages/stage-2/layouts';

export const ALL_LEVEL_LAYOUTS = [
    ...STAGE1_LAYOUTS,
    ...STAGE2_LAYOUTS,
];

export const MAX_LEVELS = ALL_LEVEL_LAYOUTS.length;

export const createBricksForStage = (stage: number): Brick[] => {
    const bricks: Brick[] = [];
    const layout = ALL_LEVEL_LAYOUTS[stage - 1];
    if (!layout) return [];

    let currentId = 0;
    layout.forEach((row, r) => {
        const cols = row.length;
        const totalBrickWidth = cols * (BRICK_WIDTH + BRICK_GAP) - BRICK_GAP;
        const startX = (GAME_WIDTH - totalBrickWidth) / 2;

        row.forEach((brickTypeValue, c) => {
            if (brickTypeValue === undefined || brickTypeValue === null) return;
            
            const brickType = brickTypeValue as BrickType;
            const brickProps = BRICK_PROPERTIES[brickType];
            
            if (brickType === BrickType.Boss) {
                bricks.push({
                    id: currentId++,
                    x: GAME_WIDTH / 2 - 90,
                    y: 20,
                    width: 180,
                    height: 50,
                    type: BrickType.Boss,
                    hp: brickProps.maxHp,
                    maxHp: brickProps.maxHp,
                    vx: BOSS_MOVE_SPEED,
                    lastAttackTime: Date.now()
                });
            } else if (brickType === BrickType.ArchmageBoss) {
                 bricks.push({
                    id: currentId++,
                    x: GAME_WIDTH / 2 - 90,
                    y: 20,
                    width: 180,
                    height: 50,
                    type: BrickType.ArchmageBoss,
                    hp: brickProps.maxHp,
                    maxHp: brickProps.maxHp,
                    lastAttackTime: Date.now(),
                    phase: 1,
                    currentElementalAttack: 'fire'
                });
            } else {
                 bricks.push({
                    id: currentId++,
                    x: startX + c * (BRICK_WIDTH + BRICK_GAP),
                    y: r * (BRICK_HEIGHT + BRICK_GAP) + 80,
                    width: BRICK_WIDTH,
                    height: BRICK_HEIGHT,
                    type: brickType,
                    hp: brickProps.maxHp,
                    maxHp: brickProps.maxHp,
                    vx: brickType === BrickType.Soldier ? 1 : 0,
                });
            }
        });
    });

    // Apply Rune brick shields
    const runeBricks = bricks.filter(b => b.type === BrickType.Rune);
    if (runeBricks.length > 0) {
        const SHIELD_HP = 1;
        const shieldedBrickIds = new Set<number>();

        for (const runeBrick of runeBricks) {
            for (const targetBrick of bricks) {
                if (targetBrick.id === runeBrick.id) continue;
                if (targetBrick.type === BrickType.Rune || targetBrick.type === BrickType.Boss || targetBrick.type === BrickType.ArchmageBoss) continue;

                // Check for horizontal adjacency
                const isAdjacentHorizontally = Math.abs(runeBrick.y - targetBrick.y) < 5 && Math.abs(runeBrick.x - targetBrick.x) < (BRICK_WIDTH + BRICK_GAP + 5);
                // Check for vertical adjacency
                const isAdjacentVertically = Math.abs(runeBrick.x - targetBrick.x) < 5 && Math.abs(runeBrick.y - targetBrick.y) < (BRICK_HEIGHT + BRICK_GAP + 5);

                if (isAdjacentHorizontally || isAdjacentVertically) {
                    shieldedBrickIds.add(targetBrick.id);
                }
            }
        }
        
        bricks.forEach(brick => {
            if (shieldedBrickIds.has(brick.id)) {
                brick.shieldHp = SHIELD_HP;
            }
        });
    }

    return bricks;
};