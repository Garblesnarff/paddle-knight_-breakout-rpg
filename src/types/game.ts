// Core shared types for engine contracts and systems

export interface Vector2D {
  x: number;
  y: number;
}

export interface Rectangle {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface PhysicsEntity {
  id: string;
  position: Vector2D;
  bounds: Rectangle;
  velocity?: Vector2D;
  mass?: number;
}

export interface Collision {
  entityAId: string;
  entityBId: string;
  contactPoint?: Vector2D;
  normal?: Vector2D;
}

export interface PhysicsUpdate {
  // Describe changes produced by a physics step
  movedEntities?: Array<{ id: string; position: Vector2D; velocity?: Vector2D }>;
  collisions?: Collision[];
}

export interface SystemUpdate {
  // Generic system update envelope; concrete systems can extend via declaration merging
  events?: string[];
}

export interface GameState {
  // Minimal placeholder; will be expanded during migration
  timeMs: number;
  entitiesById: Record<string, unknown>;
}


