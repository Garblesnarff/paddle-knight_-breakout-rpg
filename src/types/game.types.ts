export enum GameStatus {
  Start,
  StageSelect,
  Playing,
  Paused,
  GameOver,
  Victory,
  VictoryScreen,
  SkillTree,
  Targeting,
}

export enum BrickType {
  // Stage 1
  Grunt,
  Soldier,
  Archer,
  Mage,
  Tank,
  Chaos,
  Boss,
  // Stage 2
  Apprentice,
  Fire,
  Ice,
  Lightning,
  Mirror,
  Rune,
  ArchmageBoss,
}

export interface Brick {
  id: number;
  x: number;
  y: number;
  hp: number;
  maxHp: number;
  type: BrickType;
  width: number;
  height: number;
  vx?: number;
  lastAttackTime?: number;
  lastMissileTime?: number;
  lastSummonTime?: number;
  shieldHp?: number;
  phase?: number;
  lastElementalStormTime?: number;
  currentElementalAttack?: 'fire' | 'ice' | 'lightning';
  isClone?: boolean;
  realBossId?: number;
  lastMirrorImageTime?: number;
  lastManaBurnTime?: number;
  lastArcaneOverloadTime?: number;
  lastChaosMagicTime?: number;
  isFinalGambit?: boolean;
}

export interface Ball {
  id: number;
  x: number;
  y: number;
  vx: number;
  vy: number;
  size: number;
  damage: number;
  slowedUntil?: number;
  isSpikeSlowedUntil?: number;
}

export interface PlayerStats {
  power: number;
  defense: number;
  agility: number;
  luck: number;
  wisdom: number;
  vitality: number;
}

export interface Skill {
  id: string;
  name: string;
  cooldown: number;
  lastUsed: number;
  duration?: number;
  activeUntil?: number;
  charges?: number;
}

export interface Projectile {
  id: number;
  x: number;
  y: number;
  vy: number;
  size: number;
}

export interface HomingProjectile {
  id: number;
  x: number;
  y: number;
  vx: number;
  vy: number;
  size: number;
}

export interface Explosion {
  id: number;
  x: number;
  y: number;
  radius: number;
  duration: number;
  createdAt: number;
}

export interface ArcaneOrb {
  id: number;
  x: number;
  y: number;
  vx: number;
  vy: number;
  radius: number;
  damage: number;
}

export interface ElementalBeam {
  id: number;
  x1: number;
  y1: number;
  x2: number;
  y2: number;
  createdAt: number;
  duration: number;
}

export interface FireRainZone {
  id: number;
  x: number;
  y: number;
  radius: number;
  createdAt: number;
  duration: number;
}

export interface IceSpikeField {
  id: number;
  x: number;
  y: number;
  width: number;
  height: number;
  createdAt: number;
  duration: number;
}

export interface LightningStrike {
  id: number;
  x: number;
  y: number;
  width: number;
  height: number;
  createdAt: number;
  warningDuration: number;
  strikeDuration: number;
}

export interface ArcaneOverloadRing {
  id: number;
  x: number;
  y: number;
  createdAt: number;
  duration: number;
  maxRadius: number;
}

export interface FinalGambitBeam {
  id: number;
  x: number;
  width: number;
  createdAt: number;
  warningDuration: number;
  strikeDuration: number;
}

export interface BallHistoryEntry {
  x: number;
  y: number;
  vx: number;
  vy: number;
  timestamp: number;
}

export interface RunicEmpowermentBuffs {
  haste: boolean;
  power: boolean;
  shield: boolean;
}

export interface Cosmetics {
  paddleEffect?: string;
  ballEffect?: string;
}
