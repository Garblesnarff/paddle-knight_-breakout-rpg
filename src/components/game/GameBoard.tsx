import React from 'react';
import {
  Ball as BallType,
  Brick as BrickType,
  Projectile,
  Explosion,
  RunicEmpowermentBuffs,
  ArcaneOrb,
  ElementalBeam,
  HomingProjectile,
  FireRainZone,
  IceSpikeField,
  LightningStrike,
  ArcaneOverloadRing,
  FinalGambitBeam,
  Cosmetics,
  Skill,
} from '../../types/game.types';
import { GAME_WIDTH, GAME_HEIGHT, PADDLE_Y } from '../../constants/game.constants';
import { Ball } from './Ball';
import { Brick } from './Brick';
import { Paddle } from './Paddle';

interface GameBoardProps {
  balls: BallType[];
  bricks: BrickType[];
  projectiles: Projectile[];
  homingProjectiles: HomingProjectile[];
  explosions: Explosion[];
  arcaneOrbs: ArcaneOrb[];
  elementalBeams: ElementalBeam[];
  fireRainZones: FireRainZone[];
  iceSpikeFields: IceSpikeField[];
  lightningStrikes: LightningStrike[];
  arcaneOverloadRings: ArcaneOverloadRing[];
  finalGambitBeams: FinalGambitBeam[];
  paddleX: number;
  paddleWidth: number;
  cosmetics: Cosmetics;
  activeBuffs: RunicEmpowermentBuffs;
  skills: Record<string, Skill>;
  isBallLaunched: boolean;
  gameStatus: string;
  levelUpMessage: string | null;
  mousePos: { x: number; y: number };
}

export const GameBoard: React.FC<GameBoardProps> = ({
  balls,
  bricks,
  projectiles,
  homingProjectiles,
  explosions,
  arcaneOrbs,
  elementalBeams,
  fireRainZones,
  iceSpikeFields,
  lightningStrikes,
  arcaneOverloadRings,
  finalGambitBeams,
  paddleX,
  paddleWidth,
  cosmetics,
  activeBuffs,
  skills,
  isBallLaunched,
  gameStatus,
  levelUpMessage,
  mousePos,
}) => {
  return (
    <div
      className="relative shadow-inner cursor-pointer bg-black/50 border-2 border-slate-900"
      style={{ width: GAME_WIDTH, height: GAME_HEIGHT }}
    >
      {fireRainZones.map((zone) => {
        const progress = (Date.now() - zone.createdAt) / zone.duration;
        return (
          <div
            key={zone.id}
            className="absolute rounded-full bg-red-600/50 border-2 border-red-400 animate-pulse"
            style={{
              left: zone.x - zone.radius,
              top: zone.y - zone.radius,
              width: zone.radius * 2,
              height: zone.radius * 2,
              opacity: 0.8 - progress * 0.5,
              zIndex: 5,
            }}
          />
        );
      })}
      {iceSpikeFields.map((field) => {
        const progress = (Date.now() - field.createdAt) / field.duration;
        return (
          <div
            key={field.id}
            className="absolute rounded-lg bg-cyan-500/30 backdrop-blur-sm border-2 border-cyan-300"
            style={{
              left: field.x,
              top: field.y,
              width: field.width,
              height: field.height,
              opacity: 0.6 - progress * 0.4,
              boxShadow: 'inset 0 0 15px rgba(207, 250, 254, 0.5)',
              zIndex: 5,
            }}
          >
            <div
              className="absolute w-full h-full bg-no-repeat opacity-50"
              style={{
                backgroundImage: `url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 10 20'%3e%3cpath d='M5 0 L0 20 L10 20 Z' fill='rgba(255,255,255,0.2)'/%3e%3c/svg%3e")`,
                backgroundSize: '20px 20px',
              }}
            ></div>
          </div>
        );
      })}
      {lightningStrikes.map((strike) => {
        const now = Date.now();
        const warningProgress = (now - strike.createdAt) / strike.warningDuration;
        const isWarning = warningProgress < 1;
        const strikeProgress = (now - (strike.createdAt + strike.warningDuration)) / strike.strikeDuration;
        const isStriking = !isWarning && strikeProgress < 1;

        if (isWarning) {
          return (
            <div
              key={strike.id}
              className="absolute border-4 border-dashed border-yellow-400/80 rounded-lg animate-pulse"
              style={{
                left: strike.x,
                top: strike.y,
                width: strike.width,
                height: strike.height,
                zIndex: 5,
              }}
            />
          );
        }
        if (isStriking) {
          return (
            <div
              key={strike.id}
              className="absolute bg-yellow-300 rounded-lg"
              style={{
                left: strike.x,
                top: strike.y,
                width: strike.width,
                height: strike.height,
                boxShadow: '0 0 30px 15px #fef08a',
                opacity: 1 - strikeProgress,
                zIndex: 5,
              }}
            />
          );
        }
        return null;
      })}
      {arcaneOverloadRings.map((ring) => {
        const progress = (Date.now() - ring.createdAt) / ring.duration;
        const currentRadius = ring.maxRadius * progress;
        return (
          <div
            key={ring.id}
            className="absolute rounded-full border-2 border-purple-400"
            style={{
              left: ring.x - currentRadius,
              top: ring.y - currentRadius,
              width: currentRadius * 2,
              height: currentRadius * 2,
              opacity: 1 - progress,
              zIndex: 5,
            }}
          />
        );
      })}
      {finalGambitBeams.map((beam) => {
        const now = Date.now();
        const warningProgress = (now - beam.createdAt) / beam.warningDuration;
        const isWarning = warningProgress < 1;
        const strikeProgress = (now - (beam.createdAt + beam.warningDuration)) / beam.strikeDuration;
        const isStriking = !isWarning && strikeProgress < 1;

        if (isWarning) {
          return (
            <div
              key={beam.id}
              className="absolute border-y-4 border-x-8 border-dashed border-red-500/80 animate-pulse"
              style={{
                left: beam.x,
                top: 0,
                width: beam.width,
                height: GAME_HEIGHT,
                zIndex: 5,
                opacity: (Math.sin(now / 100) + 1) / 2,
              }}
            />
          );
        }
        if (isStriking) {
          return (
            <div
              key={beam.id}
              className="absolute bg-gradient-to-t from-red-600 via-pink-500 to-purple-800"
              style={{
                left: beam.x,
                top: 0,
                width: beam.width,
                height: GAME_HEIGHT,
                boxShadow: '0 0 60px 30px #ef4444',
                opacity: 0.9,
                zIndex: 5,
                animation: 'pulse 0.5s cubic-bezier(0.4, 0, 0.6, 1) infinite',
              }}
            />
          );
        }
        return null;
      })}

      {bricks.map((brick) => (
        <Brick key={brick.id} brick={brick} />
      ))}
      {explosions.map((exp) => {
        const progress = (Date.now() - exp.createdAt) / exp.duration;
        return (
          <div
            key={exp.id}
            className="absolute rounded-full bg-orange-500/70"
            style={{
              left: exp.x - exp.radius,
              top: exp.y - exp.radius,
              width: exp.radius * 2,
              height: exp.radius * 2,
              transform: `scale(${progress})`,
              opacity: 1 - progress,
            }}
          />
        );
      })}
      <svg className="absolute top-0 left-0 w-full h-full pointer-events-none z-10">
        {elementalBeams.map((beam) => {
          const progress = (Date.now() - beam.createdAt) / beam.duration;
          return (
            <line
              key={beam.id}
              x1={beam.x1}
              y1={beam.y1}
              x2={beam.x2}
              y2={beam.y2}
              stroke="#fbbf24"
              strokeWidth="5"
              strokeLinecap="round"
              style={{ opacity: 1 - progress, filter: 'drop-shadow(0 0 8px #f59e0b)' }}
            />
          );
        })}
      </svg>
      {arcaneOrbs.map((orb) => (
        <div
          key={orb.id}
          className="absolute rounded-full bg-gradient-to-br from-purple-400 to-fuchsia-600 shadow-xl border-2 border-purple-200/50"
          style={{
            left: orb.x,
            top: orb.y,
            width: orb.radius * 2,
            height: orb.radius * 2,
            boxShadow: '0 0 15px 5px #a855f7',
            animation: 'pulse 1.5s cubic-bezier(0.4, 0, 0.6, 1) infinite',
          }}
        ></div>
      ))}
      {homingProjectiles.map((p) => (
        <div
          key={p.id}
          className="absolute rounded-full bg-gradient-to-br from-purple-500 to-fuchsia-700 shadow-lg"
          style={{
            left: p.x,
            top: p.y,
            width: p.size,
            height: p.size,
            boxShadow: '0 0 8px 3px #d946ef',
          }}
        ></div>
      ))}
      {balls.map((ball) => (
        <Ball key={ball.id} ball={ball} cosmetics={cosmetics} activeBuffs={activeBuffs} />
      ))}
      {projectiles.map((p) => (
        <div
          key={p.id}
          className={`absolute rounded ${
            p.size > 8 ? 'bg-orange-500 shadow-lg shadow-orange-500/50' : 'bg-pink-500 rounded-full'
          }`}
          style={{ left: p.x, top: p.y, width: p.size, height: p.size * 2 }}
        ></div>
      ))}
      <Paddle paddleX={paddleX} paddleWidth={paddleWidth} />
      {skills.barrier.activeUntil && Date.now() < skills.barrier.activeUntil && (
        <div
          className="absolute bg-cyan-400/30 border-t-2 border-cyan-200 animate-pulse"
          style={{ left: 0, top: PADDLE_Y - 10, width: GAME_WIDTH, height: 10 }}
        ></div>
      )}

      {!isBallLaunched && gameStatus === 'Playing' && (
        <div
          className="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 text-2xl font-bold text-white/80 animate-pulse pointer-events-none font-medieval tracking-wider"
          style={{ textShadow: '2px 2px 4px #000' }}
        >
          Click to Launch
        </div>
      )}

      {levelUpMessage && (
        <div className="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 text-5xl font-medieval text-yellow-300 animate-bounce drop-shadow-lg z-50">
          {levelUpMessage}
        </div>
      )}

      {gameStatus === 'Targeting' && (
        <div className="pointer-events-none absolute inset-0 z-20 backdrop-blur-sm bg-black/30">
          <div
            className="absolute transition-transform duration-150"
            style={{ transform: `translate(${mousePos.x}px, ${mousePos.y}px)` }}
          >
            <div className="w-8 h-8 -m-4 rounded-full border-2 border-dashed border-cyan-300 animate-spin-slow"></div>
            <div className="absolute top-0 left-0 w-px h-4 -mt-4 bg-cyan-300"></div>
            <div className="absolute bottom-0 left-0 w-px h-4 -mb-4 bg-cyan-300"></div>
            <div className="absolute top-0 left-0 h-px w-4 -ml-4 bg-cyan-300"></div>
            <div className="absolute top-0 right-0 h-px w-4 -mr-4 bg-cyan-300"></div>
          </div>
          <svg className="absolute top-0 left-0 w-full h-full">
            <line
              x1={paddleX + paddleWidth / 2}
              y1={PADDLE_Y}
              x2={mousePos.x}
              y2={mousePos.y}
              stroke="rgba(103, 232, 249, 0.5)"
              strokeWidth="2"
              strokeDasharray="5, 5"
            />
          </svg>
          <div
            className="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-24 text-2xl font-bold text-white/90 font-medieval tracking-wider"
            style={{ textShadow: '2px 2px 4px #000' }}
          >
            Aim and Click to Fire Orb
          </div>
        </div>
      )}
    </div>
  );
};
