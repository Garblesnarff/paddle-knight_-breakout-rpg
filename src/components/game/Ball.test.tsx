import React from 'react';
import { render } from '@testing-library/react';
import { Ball } from './Ball';
import { Ball as BallType, Cosmetics, RunicEmpowermentBuffs } from '../../types/game.types';

test('renders Ball component', () => {
  const ball: BallType = {
    id: 1,
    x: 100,
    y: 100,
    vx: 5,
    vy: 5,
    size: 10,
    damage: 1,
  };
  const cosmetics: Cosmetics = { ballEffect: 'none' };
  const activeBuffs: RunicEmpowermentBuffs = { haste: false, power: false, shield: false };

  const { container } = render(<Ball ball={ball} cosmetics={cosmetics} activeBuffs={activeBuffs} />);
  expect(container.firstChild).toMatchSnapshot();
});
