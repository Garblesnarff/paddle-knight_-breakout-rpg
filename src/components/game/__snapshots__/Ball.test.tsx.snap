// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders Ball component 1`] = `
<div
  class="absolute rounded-full bg-gradient-to-br from-yellow-300 to-orange-500 shadow-xl border-2 border-yellow-100/50"
  style="left: 100px; top: 100px; width: 20px; height: 20px;"
/>
`;
